### ISSUE TITLE
rfc: Ingress support

### ISSUE THREAD

**Issue by grampelberg** [linkerd/linkerd2#629]
# User Stories

- As an operator, I would like to expose a set of HTTP based services to the outside world, so that the public can use my company's applications.
- As an operator, I would like to expose a set of gRPC based services to the outside world, so that applications on my corporate network (outside of a single k8s cluster) can use my cluster's applications. (Note: there's an identical story to this with TCP services for things like databases and message queues)
- As a developer, I would like to expose what I'm working on to the outside world, so that I can test functionality manually.
- As someone brand new to kubernetes, I would like to access what I just deployed, so that I can see the fruition of my work.

# Rationale

As someone already running Conduit, I'd love to remove an extra layer of abstraction (ex. nginx-ingress) and reap all the benefits I've become accustomed to.

There are some awesome tutorial flows that can come out of this (in particular by using the amazing instrumentation and visualization already in conduit). 

# Open Questions

## Technical

- For non-HTTP use cases, can Conduit leverage services with `type=LoadBalancer`?
- Is it a requirement to support the kubernetes `Ingress` resource?
- I threw the gRPC user story in there, because it'll definitely come up. As someone who doesn't know as much as he should about gRPC, is it even possible to leverage ingress as it is today to work with gRPC? 
    - Looks like it is (https://www.nginx.com/blog/nginx-1-13-10-grpc/), but you'd need some kinda of extra helpers to make it usable with ingress.

## Process

- I'm sure that I'm missing some user stories (and personas). Anyone have some suggestions?
- This could be an *enormous* issue. What are some good ways to scope it down, leverage what is awesome in Conduit and provide the biggest bang for the buck?

**Comment by grampelberg** [comment]
- @olix0r how would you like to keep this moving forward? Feels like something fun to dig into =)
- I didn't find a common template for larger issues. Any suggestions on things I missed, things to remove?
- I'm introducing a couple personas and user stories. Does this format work for folks?

**Comment by olix0r** [comment]
@grampelberg thanks for opening this issue and for introducing some personas! we're still working out the best way to structure these larger design issues.

As I've been thinking about ingress previously, (1) i've hoped to tie this to [conduit routing policy](#475) (which is in progress but has been pushed down the stack as other related issues have arisen) and (2) we've wanted to fit this into the larger TLS story (which is also in progress at the moment).

One question I have for an initial implementation is: is it a requirement to support the kubernetes `Ingress` type?  My understanding is that this API is woefully unspecified (such that all implementations have notable differences). if there were an alternate, conduit-specific primitive to manage ingress rules, would that be sufficient?  Or perhaps we need a way to synthesize conduit's route rules based on k8s ingress resources?

**Comment by grampelberg** [comment]
Hmm, added a couple new questions to the outstanding once (the TCP one is particularly interesting to me).

As for `Ingress`, it only works for HTTP today, but does provide:

- TLS
- Routing

To work correctly (tm) with cloud providers, some of the ingress solutions (see ingress-nginx) create their own cluster service (`type=LoadBalancer`) to get an external address (which they then add to the ingress).

The biggest upside that I see with the existing ingress resource is the magical `cert-manager` integration for TLS. It just worked for me, and I'd hate to have to do that again.

The biggest downside to `Ingress` is that the current standard is adding annotations for any desired configuration. ingress-nginx has over 50 right now!

It doesn't help that visibility into what's happening when you're a beginning user feels like a black box. If something doesn't work, it is super difficult to debug and understand how all the pieces fit together.

# Example Integrations

Note that all three of these have some basic concept of a 1:1 relationship between an application and the resource object. There's not a single, monolithic configuration across everything on the cluster. Each resource would be independent. 

This becomes super important when you look at solutions like helm. It becomes difficult to provide a cluster level ingress controller and modify that configuration across multiple charts arbitrarily.

## Status Quo

- Keep to L7/HTTP for now
- Use annotations for configuration
- Rely on services of `type=LoadBalancer` for cloud provider integration.

## Top Level CRD

- Support TCP/L4 + HTTP/L7 immediately.
- Configuration is easy to introspect instead of being somewhat magical.
- Debugging can be improved via. events attached to the CRD.

## Resource level annotations

I've really enjoyed prometheus' resource level annotations. It sure is easy to just slap an annotation on a deployment and have it just start working.

# Possible Delivery Path

I'm wondering if there's a way to do all of these (at least the top two) and keep from boiling the ocean. Maybe keeping the status quo to get an HTTP solution going and then implement a CRD as the configuration options explode.

**Comment by lalomartins** [comment]
If I can add 2¢.

I've been banging my head against the grpc ingress issue for months and, increasingly, I'm unconvinced the Ingress resource was a good idea in the first place. What others said in this issue are symptoms: poorly specified, lots of annotations. It's trying to sweep together wildly different solutions with different configuration models, and that ends up in a lower-common-denominator interface or lots of extensions or both.

In my ideal world, a service mesh like Conduit would implement ingress (lowercase i) using whatever interface makes sense to its internal model. (Personally I find the per-resource annotation idea brilliant!) Maybe I'd still need to do something specific in order to get an external IP, maybe not. Then, *maybe*, optionally, on top of that someone can write an Ingress controller to give easy access to a limited subset of the functionality.

Note, I fully expect things to change. In maybe a year or two Ingress should mature. Maybe part of the difficulty is that the problem Ingress is trying to solve is itself still evolving and being better understood. But like with many things on the web, having more implementations will help that evolution along more likely than hinder.

tl;dr as someone who wants to expose both grpc services and static web resources from a k8s cluster running Conduit, I'd prefer if Conduit could take care of ingress for me (with all the goodies it provides), and using Ingress resources is not a requirement, I'm sure the Conduit team can come up with a more effective interface.

**Comment by grampelberg** [comment]
@lalomartins that's some awesome feedback. There's definitely a short term solution and a long term solution taht we should probably be working on in parallel.

**Comment by dlaidlaw** [comment]
This sounds like it would become an API Gateway, and maybe it can be implemented like Ambassador is implemented. [Ambassador](https://www.getambassador.io/) is a separate project that leverages Envoy for its service mesh features. Ambassador uses annotations on the Service object to configure itself, rather than Ingress.

Another ingress controller leveraging Envoy is the [Heptio Contour](https://github.com/heptio/contour) project. This is not an API Gateway, but it is similar to what is outlined in the posts above. Contour uses Custom Resource Definitions for its [IngressRoute](https://github.com/heptio/contour/blob/master/docs/ingressroute.md) definitions.

I mention these simply as examples of what is already out there, not as a recommendation for either. Also, I am not involved in either of those projects in any way.

**Comment by grampelberg** [comment]
For now, we've decided to integrate with ingress controllers specifically to leverage the awesome CNCF community and let the existing ingress controllers do what they do best.
